### ISSUE TITLE
Taglib 2 Release

### ISSUE THREAD

**Issue by v00d00** [taglib/taglib#837]
Do you have any idea when the next release will be?

**Comment by CyberSinh** [comment]
Hello,

Does anyone work to release a version 2 of taglib in the coming months? 
Indeed, the activity of the Git repository seems very weak in recent months.

Thanks.

**Comment by sbooth** [comment]
I would love to make 2.0 the primary branch but it would also be possible to do a 2.0 release and maintain the 1.0 branch for compatibility.  I know some projects are using TagLib v2 and it's quite stable, but I don't believe it is widely adopted.  I'm not sure how important it is to maintain a binary compatible version.

**Comment by CyberSinh** [comment]
It is a good idea to maintain a 1.x branch with only bug fixes (to limit the workload in v1) and develop all new features in a v2 branch.
But the v2 the will never be the most used version as long as it will not be the main branch of development (master). That's why I think it is time to switch to v2 as primary branch.
Keeping the binary compatibility between major version should not be an objective because it would limit the possible innovation for the v2.

**Comment by v00d00** [comment]
Yeah exactly, people cant really use taglib2 at all yet, even with all the good stuff.

Tag a release to start the taglib2 journey!

**Comment by a17r** [comment]
Regardless of 2, are there plans for a new taglib-1.11.x release in the near future? clementine in master has begun to depend on unreleased taglib...

**Comment by scotchi** [comment]
Just for the folks wondering -- for the last couple of releases @TsudaKageyu handled those, however, I haven't heard anything from him in the project for more than a year, so my assumption is that he's moved on.

For a 2.x release that leaves us in something of an odd spot, since the push for a 2.x branch mostly came from @TsudaKageyu and @lalinsky, but neither of them are active in the project at the moment.

In the shorter term, it's likely that there will be another 1.x bugfix release that I'll put together if nobody else jumps forward to do.  (On the other hand, I've got a second kid that was just born, so time is quite tight at the moment.)

**Comment by sbooth** [comment]
I could probably roll a 1.x bug fix release in the next week or so.

I also am a proponent of pushing 2.0 out in the wild. I think there are one or two PRs that need to be integrated first.

**Comment by v00d00** [comment]
Great news! Bring on taglib 2.0 and 1.12! :+1:

**Comment by jonaski** [comment]
Any news on this?
@sbooth and @scotchi 
Is there anything more that needs to be done for a new 1.x release?

**Comment by vtorri** [comment]
hello

I am also interested in a new 1.12 release.

**Comment by ArielRosenfeld** [comment]
Could you please release a taglib v1 release so packagers can have an up to date version with fixes.

packagers dont like to use git versions and simply get patches for the 1.11.1 version if bugs are reported in the distribution bug system.

it has been almost two years since the last release and software that depends on taglib tend to package a git version and not use the system taglib.

Having a release will make it easier to contribute, as it will created a none patched version to target for bugs and features.

if you think the version isnt ready for release yet consider a RC or Beta release, it will at least unify what distributions use.

Releasing a timed release once or twice a year also attracts developers as git are only visible if somebody bothers to in the development website, an update showing up in distributions shows the project isnt dead for a larger community.

**Comment by scotchi** [comment]
This wasn't an active decision that was made -- the TagLib maintainer disappeared in that timeframe, and I (as the original author) didn't have the time to step up and handle releases then (new baby at home, lots of work).

I have a little bit of breathing room again, and in the last week I've been going through and reviewing outstanding patches again, with an eye on getting a release out in the next few weeks.  This will notably be v1 only.  I'm still not sure what the future of the v2 branch will be as activity there was primarily from folks that are no longer active in the project.

**Comment by equeim** [comment]
Thank you for your work! What about new features introduced in taglib2? I'm particularly interested in pictures API.

**Comment by scotchi** [comment]
I don't have a good answer for that at the moment.  The truth is that I was against the creation of the Taglib2 branch, and the folks that were pushing for it have since left the project.

**Comment by sbooth** [comment]
Personally I'm an advocate for the TagLib2 branch but it would probably be best for the project long-term not to have two parallel branches that are so similar without a clear plan to transition from 1 to 2. Currently 2 supports a better pictures API and EBML.

**Comment by scotchi** [comment]
From the 1.x branch, there's now a new tag available for testing:

https://github.com/taglib/taglib/tree/v1.12-beta-1

**Comment by whyman** [comment]
@scotchi Any chance of a 1.12.x release in 2020? :grin:

**Comment by hyperquantum** [comment]
As a developer who uses Taglib 1.x, I have no idea what Taglib 2 would offer me. Does it only fix some basic flaws from version 1 that cannot be fixed in 1.x because of backward compatibility? Does it have an API that is easier to use? Does it offer some significant extra functionality that cannot be found in version 1.x? Those things should be documented somewhere. Are they? I can't seem to find it anywhere.

If nobody knows why Taglib 2 is better than version 1 then there is no incentive for developers to consider migrating, and we will just be stuck with version 1.x and its backward compatibility kludges forever.

**Comment by sjlongland** [comment]
Are there plans to do any new releases of this library?

Reason I ask, I have an idea for a media server system built around the Matroska file format but with support for other formats as well -- the plan being that you'd be able to import and export media in some arbitrary format, and Matroska would be used to store it in some common format losslessly (i.e. MP3 would be stored as MP3, AAC as AAC, FLAC as FLAC, but all packaged up as `mka`).  Did a look around, there aren't too many alternatives that let me do this.  `ogm` is dead, `avi` is a heavily-constrained 90s relic, most are patent encumbered.  Google (via WebM) and Microsoft (directly) seem to be throwing their collective weight behind Matroska.

Matroska is not supported in Taglib v1, which means all software built to use Taglib similarly, are locked to whatever Taglib v1 supports.

This situation is not ideal.  Software distribution repositories (e.g. Linux APT/DNF/etc archives; BSD ports, etc) will generally shy away from snapshotting unreleased package repositories unless they have no other choice.  Given this, projects that depend on the newer features wind up either shipping their own "bundled" taglib (bad for security/efficiency) or re-invent the wheel (worse).

@hyperquantum makes a very valid point: why would an established project adopt something that has no apparent benefit?

I've also had a look at the mailing list archive.  Beyond the announcement of [v1.12](https://mail.kde.org/pipermail/taglib-devel/2021-February/003012.html), activity seems very quiet.

Is there some sort of road-map that points out what's happening with the project in the mid-long term regarding releases?

**Comment by scotchi** [comment]
It's not clear to me what will happen with the taglib2 branch.  I was opposed to it being created (for me it never made sense to have a divergent branch with no release plan), and the people that created it don't work on TagLib anymore.  It's possible @ufleisch has some interest in it, but if so, I can't remember him broadcasting that.

As mentioned in #967, the thing keeping out Matroska support isn't a release of the taglib2 branch, it's the code being clean, complete and passing unit tests.

In general TagLib is a mature library.  I think the general goal is that when there's enough stuff that's made it in, or are some critical bug fixes that someone rolls up a new release, but there's not anyone that's actively trying to take TagLib in new directions.  TagLib is 20 years old, and mostly does what it's supposed to.

**Comment by sjlongland** [comment]
Right, so basically some of these "in development" features need to be polished up (design issues resolved, unit tests added, etc).

Indeed, we don't need to be making new releases every fortnight, this isn't a web browser.  Is there a similar TO-DO list of sorts that some of us could pitch-in and help with?

**Comment by scotchi** [comment]
I think the closest approximation of a to-do list is the issue tracker / merge requests.  There are certainly some things things there that could use review or completion.  I don't think from myself and @ufleisch there's much of a wishlist for TagLib -- again, mostly just because it does what it's supposed to.  (e.g. in the case of Matroshka, I've never had much interest in supporting formats that are primarily video containers, since TagLib is  an audio meta data library, but I'm also not opposed to looking them over if someone gets them into a usable state)

**Comment by ufleisch** [comment]
I am afraid that switching to the taglib2 branch would introduce some bugs because this branch lived his own life for a long time and it did not receive all fixes which were applied to the master branch. Up to some point in time, the commits to the master branch were backported, but since the taglib2 branch is certainly used a lot less than the master branch, it is not tested to an extent that would give great confidence.

So I personally would not advice to switch to this branch in the future. If we decide to make a version 2.0 which is binary incompatible, I would rather make this switch in the master branch and from that point in time work solely on this 2.0 version while the 1.x release branch would only receive important bug fixes when needed. Probably, some code from the taglib2 branch could be useful to integrate in an incremental way rather than as a big bang switch. It would probably also be a good thing to keep such a 2.0 release source compatible with the 1.x releases so that projects would adopt it.

Concerning the Matroska support, I would appreciate such a feature and would suggest the following procedure: Have a look at the code in the corresponding merge request #967 and compare it with the MKV code from the taglib2 branch, it might be similar. Look at the comment I made a year ago, these issue would have to be solved. You would have to judge whether the code is qualitatively sufficient to build on. As the information in the MKV tags with target type and nested tags is quite special compared with other tag formats, a format specific API for this information would be needed with a good mapping to TagLib properties, so that all information in such a tag could be listed with `tagreader` and other software using properties. The mapping to the standard TagLib API (title, artist, ...) should be chosen in a way to be compatible with the information shown in other software such as Handbrake and VLC.

Such an implementation would also be useful for your project before it is included in TagLib because it is easily possible to use the standard TagLib and extend in with additional formats in your application code.

**Comment by scotchi** [comment]
> So I personally would not advice to switch to this branch in the future. If we decide to make a version 2.0 which is binary incompatible, I would rather make this switch in the master branch and from that point in time work solely on this 2.0 version while the 1.x release branch would only receive important bug fixes when needed. Probably, some code from the taglib2 branch could be useful to integrate in an incremental way rather than as a big bang switch. It would probably also be a good thing to keep such a 2.0 release source compatible with the 1.x releases so that projects would adopt it.

That's basically my feeling as well.  The master branch has always been the more mature branch, and where most development has happened.  In 2012 Lukas created the taglib2 branch, but there was never any communication or planning about moving over to that branch as the main released branch, so there have been a collection of things that are thrown into there because they're breaking changes, but without any plan to actually move over to that branch.

It looks like things were merged back that direction until about 2018, and since then it's been more or less dead.

I think the most pragmatic thing to say about the taglib2 branch is:  there appears to be nobody that is interested in working on it.  If there was, then we could debate if the changes from master could be merged into it, or if it would make sense to factor it back out from being a giant omnibus branch to a set of feature branches that implemented certain parts of "taglib2", which could eventually be merged into master, but with nobody interested in doing that work, it seems kind of a silly discussion.

What may make sense is renaming it from taglib2 to something like "experimental" to avoid the confusion that people looking at the project assume that it's going to be the basis for an upcoming release.  Again, for the 4 years that this bug report has been open, and for the 9 years that the branch has existed, there's never been a plan to move towards releasing something based on the taglib2 branch.

If that ever does happen, at that point it'd definitely make sense to stop working on 1.x for anything but backported bugfixes.  And I agree that source compatibility makes sense.  I've done so little looking at the taglib2 branch that I don't even know to what extent that's already true.  But again, it doesn't make sense to have a long debate about what should happen in a branch that nobody's working on.

**Comment by supermihi** [comment]
+1 for doing the v2 development on `master` and keeping source compatibility. Bugfix releases for v1.x can still be made via release branches and cherry-picks.
